controller/api/destination/server.go

```diff
@@ -352,6 +351,127 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
+// network, then it must have a container port that exposes `port` as a host
+// port.
+func getPod(k8sAPI *k8s.API, podIP string, port uint32, log *logging.Entry) (*corev1.Pod, error) {
+	// First we check if the address maps to a pod in the host network.
+	addr := fmt.Sprintf("%s:%d", podIP, port)
+	hostIPPods, err := getIndexedPods(k8sAPI, watcher.HostIPIndex, addr)
+	if err != nil {
+		return nil, status.Error(codes.Unknown, err.Error())
+	}
+	if len(hostIPPods) == 1 {
+		log.Debugf("found %s:%d on the host network", podIP, port)
+		return hostIPPods[0], nil
+	}
+	if len(hostIPPods) > 1 {
+		conflictingPods := []string{}
+		for _, pod := range hostIPPods {
+			conflictingPods = append(conflictingPods, fmt.Sprintf("%s:%s", pod.Namespace, pod.Name))
+		}
+		log.Warnf("found conflicting %s:%d endpoint on the host network: %s", podIP, port, strings.Join(conflictingPods, ","))
+		return nil, status.Errorf(codes.FailedPrecondition, "found %d pods with a conflicting host network endpoint %s:%d", len(hostIPPods), podIP, port)
+	}
+
+	// The address did not map to a pod in the host network, so now we check
+	// if the IP maps to a pod IP in the pod network.
+	podIPPods, err := getIndexedPods(k8sAPI, watcher.PodIPIndex, podIP)
+	if err != nil {
+		return nil, status.Error(codes.Unknown, err.Error())
+	}
+	if len(podIPPods) == 1 {
+		log.Debugf("found %s on the pod network", podIP)
+		return podIPPods[0], nil
+	}
+	if len(podIPPods) > 1 {
+		conflictingPods := []string{}
+		for _, pod := range podIPPods {
+			conflictingPods = append(conflictingPods, fmt.Sprintf("%s:%s", pod.Namespace, pod.Name))
+		}
+		log.Warnf("found conflicting %s IP on the pod network: %s", podIP, strings.Join(conflictingPods, ","))
+		return nil, status.Errorf(codes.FailedPrecondition, "found %d pods with a conflicting pod network IP %s", len(podIPPods), podIP)
+	}
+
+	log.Debugf("no pod found for %s:%d", podIP, port)
+	return nil, nil
+}
+
+func getIndexedPods(k8sAPI *k8s.API, indexName string, podIP string) ([]*corev1.Pod, error) {
+	objs, err := k8sAPI.Pod().Informer().GetIndexer().ByIndex(indexName, podIP)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting %s indexed pods: %s", indexName, err)
+	}
+	pods := make([]*corev1.Pod, 0)
+	for _, obj := range objs {
+		pod := obj.(*corev1.Pod)
+		if !podReceivingTraffic(pod) {
+			continue
+		}
+		pods = append(pods, pod)
+	}
+	return pods, nil
+}
+
+func podReceivingTraffic(pod *corev1.Pod) bool {
+	phase := pod.Status.Phase
+	podTerminated := phase == corev1.PodSucceeded || phase == corev1.PodFailed
+	podTerminating := pod.DeletionTimestamp != nil
+
+	return !podTerminating && !podTerminated
+}
+
+// podToAddressSet converts a Pod spec into a set of Addresses.
+func podToAddressSet(k8sAPI *k8s.API, pod *corev1.Pod) *watcher.AddressSet {
+	ownerKind, ownerName := k8sAPI.GetOwnerKindAndName(context.Background(), pod, true)
+	return &watcher.AddressSet{
+		Addresses: map[watcher.PodID]watcher.Address{
+			watcher.PodID{
```

Comment:
Linter is suggesting this as I review 😬:
```suggestion
			{
```